### PR TITLE
Oreon training

### PR DESCRIPTION
**Pull request by Naman-sopho**
- Add Oreon training for `Intelligence` and `Strength` attributes.
- Add buildings `Gym` and `Classroom` for this training. Each time the Oreon visits one of these buildings its corresponding attribute is increased by 10.
- Add a `DelayedNotificationSystem` so that the player is not notified of a missing building every time the Oreon's BT is run.
- Add WorldlyTooltip which currently displays all Oreon Attributes with a placeholder icon image.

Outstanding before merge:
- [x] Oreon's attribute is increased beyond the maximum limit when the difference between the max value and current value is less than 10.

### CODE REVIEW

File: src/main/java/org/terasology/taskSystem/DelayedNotificationSystem.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    @In
+    private LocalPlayer localPlayer;
+
+    /**
+     * The maximum delay time after which notification is sent again.
+     */
+    private long maxDelay = 10;
+
+    public float sendNotification(String message, EntityRef notificationMessageEntity, float lastNotification) {
+        if (time.getGameTime() - lastNotification > maxDelay || lastNotification == 0) {
```
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/DelayedNotificationSystem.java]: You should check for lastNotification == 0 first, then the subtraction > max check

File: src/main/java/org/terasology/taskSystem/DelayedNotificationSystem.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2017 MovingBlocks
```
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/DelayedNotificationSystem.java]: 2018

File: src/main/java/org/terasology/taskSystem/actions/NeedsTrainingNode.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2017 MovingBlocks
```
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/actions/NeedsTrainingNode.java]: 2018

File: src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java [hunk 4]
```diff
@@ -96,6 +96,20 @@ private void changeOreonAttributes(Actor oreon, TaskComponent taskComponent) {
                 oreonAttributeComponent.hunger = 0;
                 break;
 
+            case AssignedTaskType.Train_Strength:
+                oreonAttributeComponent.strength += 10;
+                oreonAttributeComponent.strength = oreonAttributeComponent.strength > oreonAttributeComponent.maxStrength
```
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java]: Assigning the same value doesn't make sense.  Just use an if ( str > max) { str = max; } statement.

File: src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java [hunk 5]
```diff
@@ -96,6 +96,20 @@ private void changeOreonAttributes(Actor oreon, TaskComponent taskComponent) {
...
+                        ? oreonAttributeComponent.maxStrength : oreonAttributeComponent.strength;
+                logger.info("Strength training complete, strength is now : {}", oreonAttributeComponent.strength);
+                break;
+
+            case AssignedTaskType.Train_Intelligence:
+                oreonAttributeComponent.intelligence += 10;
+                oreonAttributeComponent.intelligence = oreonAttributeComponent.intelligence > oreonAttributeComponent.maxIntelligence
```
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java]: Same as above.  use an if statement instead.

File: src/main/java/org/terasology/tooltip/TooltipClientSystem.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2017 MovingBlocks
```
Comment (mkienenb) [inline src/main/java/org/terasology/tooltip/TooltipClientSystem.java]: 2018

File: src/main/java/org/terasology/tooltip/TooltipClientSystem.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import org.terasology.world.selection.BlockSelectionComponent;
+import org.terasology.worldlyTooltipAPI.events.GetTooltipIconEvent;
+import org.terasology.worldlyTooltipAPI.events.GetTooltipNameEvent;
+
+@RegisterSystem(RegisterMode.CLIENT)
+public class TooltipClientSystem extends BaseComponentSystem {
+    @ReceiveEvent(components = OreonAttributeComponent.class)
+    public void addAttributesToTooltip(GetItemTooltip event, EntityRef entity, OreonAttributeComponent oreonAttributeComponent) {
+        event.getTooltipLines().add(new TooltipLine("Level : " + oreonAttributeComponent.currentLevel));
+        event.getTooltipLines().add(new TooltipLine("Health : " + oreonAttributeComponent.health + " / " + oreonAttributeComponent.maxStrength ));
```
Comment (mkienenb) [inline src/main/java/org/terasology/tooltip/TooltipClientSystem.java]: Shouldn't this be maxHealth, not maxStr?

File: src/main/java/org/terasology/tooltip/components/OreonTooltipComponent.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2017 MovingBlocks
```
Comment (mkienenb) [inline src/main/java/org/terasology/tooltip/components/OreonTooltipComponent.java]: 2018

File: src/main/java/org/terasology/taskSystem/DelayedNotificationSystem.java (hunk 1 above)
Comment (mkienenb) [inline src/main/java/org/terasology/taskSystem/DelayedNotificationSystem.java]: You should check for lastNotification == 0 first, then the subtraction > max check.

**Review by mkienenb — CHANGES_REQUESTED** [review]

**Review by Cervator — APPROVED** [review]
Can't find anything to add on top of existing comments, happy to merge when those are addressed :-)

Tests out well, nice work!

**Comment by Cervator** [comment]
Tweaks applied - merged! :-)

**Pull request merged**